test/test_playbooks/tasks/smart_proxy_puppet_host.yml

```diff
@@ -0,0 +1,7 @@
+---
+- name: "Ensure resolution of puppet to localhost in /etc/hosts is {{ puppet_host_state }}"
+  lineinfile:
+    path: /etc/hosts
+    line: 127.0.0.1 puppet
+    state: "{{ puppet_host_state }}"
+
```

Comment:
OH, i think, i am not going to like this. Especially, since those module tests do not need any privilege escalation at all by now. Can we please find another way. As a last resort, we specify entities, the user must provide for recording in `...vars/server.yaml`.

Reply:
I forgot to mention, that the rest looks amazing from a first look.

Reply:
The Smart Proxy uses SSL authentication, and the server side certificate is issued to <fqdn> with a SAN of 'puppet'.

Attempting to remove and add back the default Smart Proxy isn't feasible as it would require removing significant amounts of other, dependent configuration (e.g. subnets, domains, etc, etc).

The only other way I can think to do this is to change the configuration of the Foreman instance being used to record the test data prior to testing.  In this case, that would mean customizing the Forklift playbook to either add another SAN for localhost to the Smart Proxy server cert, or to add that same puppet entry to the /etc/hosts file.  Doing that would mean that the generation of these VCR cassettes will always require customization of the Foreman instance prior to recording.  Is that an acceptable solution?  Would it be acceptable to make that action and the corresponding teardown action conditional on a variable set in server.yaml that would only be used for recording VCR cassettes?

Reply:
Why do we need `puppet` to be resolvable in the first place? Can't we use `ansible_fqdn` or something?

Reply:
The FQDN of the system is what's used in the URI for the default Smart Proxy.  That URI must be globally unique, so adding another with the same URI is not possible.  The cert is valid for that FQDN and 'puppet'.  So the options here are:

* Remove the existing Smart Proxy first, which will require the teardown of most of the Foreman config (e.g. all domains, subnets, etc must be detached or removed).
* Change the Foreman setup process to include more SANs, such as localhost, and add a Smart Proxy for that URI.
* Make 'puppet' resolve to localhost and use that URI.

The 'puppet' resolution option is the smallest change set and has no environmental impacts, but it does require privilege escalation.

I see two paths forward that have a chance of being sustainable over time:
* Leave the 'puppet' resolution approach in place but ensure that the resolution stuff only runs during recording.
* Use a non-standard Foreman setup process to allow the tests to run against localhost successfully, documenting the setup process required to successfully do recording in the future.

Just let me know which way you'd like to go here.

Reply:
Oooooh right. Not sure yet, but now I know what to think about, thanks!

Reply:
Just to close the loop on this, I'm not planning to make any additional changes without further guidance.  If any of the paths forward on this become acceptable, please let me know and I'd be happy to make the changes.  Thanks!

Reply:
That's absolutely fair!

I think I agree that using the internal proxy is the cheapest and least invasive way we can get away here. But I also have to agree with Matthias that using `become` will be problematic, especially as we run our tests on travis-ci.com where we don't have the ability to get root (at least at the moment, we *could* change that).

Given that once the VCR tapes are recorded, we don't run against a real Foreman, we don't actually need any of these names resolvable during tests, only during record.

We already have tasks that are only executed when a recording happens:
https://github.com/theforeman/foreman-ansible-modules/blob/ba2c1809df34598287d483a239e919da66e5e7d5/test/test_playbooks/tasks/repository_sync.yml#L38
So I think this would be step one to make it "smarter".

I think I'm OK with accepting the fact that this particular test will need a "special" setup to be recordable (special = being able to become root).

And also see below for @ekohl's idea of using HTTP instead of HTTPS :)

Another thing that I realized, but not sure if I want to ask you to change that: You're adding `127.0.0.1 puppet`, but you do so on the host that is running the module, not on the host that is running Foreman. These might be the same machines in your setup, but e.g. in mine I record from my laptop against a VM that runs foreman, and adding that line to my laptops /etc/hosts will not help.

Now, after having written that, maybe this should be just documented at the top of `test_playbooks/smart_proxy.yml` as an requirement and not try to be smart at all? Then we could also make the smart proxy URL configurable in https://github.com/theforeman/foreman-ansible-modules/blob/master/test/test_playbooks/vars/server.yml.example and allow the users to do use whichever proxy they want? (which is basically a "yes" to your question in https://github.com/theforeman/foreman-ansible-modules/pull/314#discussion_r300125000 whether a required customization step is is acceptable).

Gee, lots of text. Hope it makes any sense to you (and not only to my weird brain ;))

Reply:
And to add to @ekohl's idea, I was able to add a `http://$FQDN:8000` proxy to a katello setup I have here, but I had to disable SELinux as our policy actually forbids connecting to port 8000 :)

Reply:
So while @ekohl is out for a shower: the installer adds the internal proxy as `https://FQDN:9090` (at least in Katello, in Foreman it's 8443?) and you can't add another one with that same url. But you can add `https://FQDN:9090/` (i.e. with a trailing slash) just fine, because we only check for string equality. That said, you can also add `https://FOREMAN.EXAMPLE.COM:9090` because we don't lowercase the URL :)

I think now we have plenty of dirty tricks to reach the local proxy w/o adjusting the OS.